### PR TITLE
Add tests for blocking gc

### PR DESCRIPTION
**Pull request by icristescu**
Adding back the tests written for the layered store.

Based on #1885

### CODE REVIEW

**Comment by icristescu** [comment]
This is ready for review. Some TODOs are left in the code, that will be removed once we add the magic `Dangling_parent_commit`.

File: test/irmin-pack/test_gc.ml [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+    let* t, c4 = commit_4 t in
+    let* t = checkout_exn t c4 in
+    let* t, c5 = commit_5 t in
+    S.reload ro_t.repo;
+    [%log.debug "Gc c3, keep c4, c5"];
+    let () = gc t c4 in
+    let () = wait_for_gc t.repo in
+    [%log.debug "RO finds c3, c4 before reload"];
+    let* () = check_3 ro_t c3 in
+    let* () = check_5 ro_t c5 in
```
Comment (tomjridge) [inline test/irmin-pack/test_gc.ml:376]: Comment says c4, but code says c5?

File: test/irmin-pack/test_gc.ml [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    S.reload ro_t.repo;
+    [%log.debug "RO finds c4 but not c3 after reload"];
+    let* () = check_5 ro_t c5 in
+    (* TODO: parent of commit is not gced. *)
```
Comment (tomjridge) [inline test/irmin-pack/test_gc.ml:380]: "RO finds c5..."?

File: test/irmin-pack/test_gc.ml [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+module Log = (val Logs.src_log src : Logs.LOG)
+
+let test_dir = "_build"
+
+let fresh_name =
+  let c = ref 0 in
+  fun () ->
+    incr c;
+    let name = Filename.concat test_dir ("test-gc" ^ string_of_int !c) in
+    name
```
Comment (tomjridge) [inline test/irmin-pack/test_gc.ml:31]: I sort-of worry that these names aren't actually fresh... what happens if we run the tests (so, lots of _build/test-gcNNNN dirs) then run the tests again? Do the dirs get reused? And if so, does the contents of the dirs mess up the tests? One option is just to use Sys.file_exists or similar, and keep incrementing the counter till we hit a fresh name. Or maybe just add a note that we expect to run the tests just once, on a _build dir with no test-gcNNNN subdirs?

File: test/irmin-pack/test_gc.ml [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+
+  let info = S.Info.empty
+
+  let gc ?unlink t commit =
+    let commit_key = S.Commit.key commit in
+    S.gc ?unlink t.repo commit_key
+
+  let wait_for_gc _ =
+    () (* TODO for testing purposes its convenient to have this function. *)
+
```
Comment (tomjridge) [inline test/irmin-pack/test_gc.ml:59]: It is not completely clear how we implement this function... but presumably it will be in lwt?

**Comment by tomjridge** [comment]
Some trivial comments; otherwise looks good to me.

File: test/irmin-pack/test_gc.ml (hunk 4 above)
Comment (Ngoguey42) [reply test/irmin-pack/test_gc.ml:59]: Good point. We can make 2 blocking functions. One direct style and another one that returns an Lwt promise

**Review by Ngoguey42 — APPROVED** [review]
Thanks

File: test/irmin-pack/test_gc.ml (hunk 4 above)
Comment (icristescu) [reply test/irmin-pack/test_gc.ml:60]: I changed it to Lwt, and we can add the direct style one later if needed

**Pull request merged**